usage/src/org/apache/cloudstack/quota/QuotaManagerImpl.java

```diff
@@ -0,0 +1,397 @@
+                    usageRecord.getStartDate(), usageRecord.getEndDate());
+            _quotaUsageDao.persist(quota_usage);
+            quotalist.add(quota_usage);
+        }
+        tariff = _quotaTariffDao.findTariffPlanByUsageType(QuotaTypes.MEMORY, usageRecord.getEndDate());
+        if (tariff != null && !tariff.getCurrencyValue().equals(0)) {
+            BigDecimal memory = new BigDecimal(serviceoffering.getRamSize());
+            onehourcostper1mb = tariff.getCurrencyValue().multiply(aggregationRatio);
+            memoryquotausage = rawusage.multiply(onehourcostper1mb).multiply(memory);
+            quota_usage = new QuotaUsageVO(usageRecord.getId(), usageRecord.getZoneId(), usageRecord.getAccountId(), usageRecord.getDomainId(), QuotaTypes.MEMORY, memoryquotausage,
+                    usageRecord.getStartDate(), usageRecord.getEndDate());
+            _quotaUsageDao.persist(quota_usage);
+            quotalist.add(quota_usage);
+        }
+        tariff = _quotaTariffDao.findTariffPlanByUsageType(QuotaTypes.RUNNING_VM, usageRecord.getEndDate());
+        if (tariff != null && !tariff.getCurrencyValue().equals(0)) {
+            onehourcostforvmusage = tariff.getCurrencyValue().multiply(aggregationRatio);
+            vmusage = rawusage.multiply(onehourcostforvmusage);
+            quota_usage = new QuotaUsageVO(usageRecord.getId(), usageRecord.getZoneId(), usageRecord.getAccountId(), usageRecord.getDomainId(), QuotaTypes.RUNNING_VM, vmusage,
+                    usageRecord.getStartDate(), usageRecord.getEndDate());
+            _quotaUsageDao.persist(quota_usage);
+            quotalist.add(quota_usage);
+        }
+
+        usageRecord.setQuotaCalculated(1);
+        _usageDao.persist(usageRecord);
+        return quotalist;
+    }
+
+    @DB
+    public QuotaUsageVO updateQuotaAllocatedVMUsage(UsageVO usageRecord, final BigDecimal aggregationRatio) {
+        QuotaUsageVO quota_usage = null;
+        QuotaTariffVO tariff = _quotaTariffDao.findTariffPlanByUsageType(QuotaTypes.ALLOCATED_VM, usageRecord.getEndDate());
+        if (tariff != null && !tariff.getCurrencyValue().equals(0)) {
+            BigDecimal vmusage;
+            BigDecimal onehourcostforvmusage;
+            onehourcostforvmusage = tariff.getCurrencyValue().multiply(aggregationRatio);
+            vmusage = new BigDecimal(usageRecord.getRawUsage()).multiply(onehourcostforvmusage);
+            quota_usage = new QuotaUsageVO(usageRecord.getId(), usageRecord.getZoneId(), usageRecord.getAccountId(), usageRecord.getDomainId(), QuotaTypes.ALLOCATED_VM, vmusage,
+                    usageRecord.getStartDate(), usageRecord.getEndDate());
+            _quotaUsageDao.persist(quota_usage);
+        }
+
+        usageRecord.setQuotaCalculated(1);
+        _usageDao.persist(usageRecord);
+        return quota_usage;
+    }
+
+    @DB
+    public QuotaUsageVO updateQuotaRaw(UsageVO usageRecord, final BigDecimal aggregationRatio, final int ruleType) {
```

Comment:
There appears to be duplication between this method and updateQuotaAllocatedVMUsage(UsageVO, BigDecimal).  Extract the common aspects into private, static methods with unit tests as necessary.